### PR TITLE
Properly parse JSON Lines output from yarn

### PR DESCRIPTION
**Pull request by jack1142**
##### SUMMARY
As [explained here](https://github.com/ansible-collections/community.general/issues/4045#issuecomment-1014585718), yarn's `--json` option returns output in a JSON Lines format and the module did not handle that properly. This fixes that.

While the issue was discovered in #4045, the issue is not necessarily limited to it (though I couldn't come up with anything that would cause it in other cases).

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
yarn

##### ADDITIONAL INFORMATION
This is not enough to address #4045 and #4048 is still needed to fix the lack of `~` expansion.

### CODE REVIEW

File: plugins/modules/packaging/language/yarn.py [hunk 1]
```diff
@@ -241,16 +241,19 @@ def list(self):
...
-
-        for dep in dependencies:
-            name, version = dep['name'].rsplit('@', 1)
-            installed.append(name)
+        for json_line in result.strip().split('\n'):
+            data = json.loads(json_line)
+            if self.globally:
+                if data['type'] == 'list':
+                    # This is a string in format: 'bins-<PACKAGE_NAME>'
+                    installed.append(data['data']['type'][5:])
```
Comment (felixfontein) [inline plugins/modules/packaging/language/yarn.py:249]: Maybe also verify that `data['data']['type'].starstwith('bins-')`?

**Review by felixfontein — COMMENTED** [review]
Thanks! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

File: changelogs/fragments/4050-properly-parse-json-lines-output-from-yarn.yaml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - yarn - fix incorrect handling of `yarn list` and `yarn global list` output that could result in fatal error (https://github.com/ansible-collections/community.general/pull/4050).
+  - yarn - fix incorrectly reported status when installing a package globally (https://github.com/ansible-collections/community.general/issues/4045, https://github.com/ansible-collections/community.general/pull/4050).
```
Comment (felixfontein) [inline changelogs/fragments/4050-properly-parse-json-lines-output-from-yarn.yaml:3]:
```suggestion
  - yarn - fix incorrect handling of ``yarn list`` and ``yarn global list`` output that could result in fatal error (https://github.com/ansible-collections/community.general/pull/4050).
  - yarn - fix incorrectly reported status when installing a package globally (https://github.com/ansible-collections/community.general/issues/4045, https://github.com/ansible-collections/community.general/pull/4050).
```

File: plugins/modules/packaging/language/yarn.py [hunk 3]
```diff
@@ -241,16 +241,19 @@ def list(self):
...
-        try:
-            dependencies = data['data']['trees']
-        except KeyError:
-            missing.append(self.name)
-            return installed, missing
-
-        for dep in dependencies:
-            name, version = dep['name'].rsplit('@', 1)
-            installed.append(name)
+        for json_line in result.strip().split('\n'):
```
Comment (felixfontein) [inline plugins/modules/packaging/language/yarn.py:244]: How about:
```suggestion
        for json_line in result.strip().splitlines():
```
That will also take care of `\r\n` line endings (if these will ever appear).

**Comment by felixfontein** [comment]
@moonrail @imjoseangel since you contributed to this module before, you might be interested in this PR as well (I don't use yarn so I cannot really test this).

File: plugins/modules/packaging/language/yarn.py (hunk 3 above)
Comment (jack1142) [reply plugins/modules/packaging/language/yarn.py:244]: Per the documentation of JSON Lines, the line separator is `\n` and trailing `\r` is not an issue as the JSON parser will just ignore it:
https://jsonlines.org/

**Comment by felixfontein** [comment]
If nobody comments in a week, I'll simply merge this.

**Comment by felixfontein** [comment]
@jack1142 thanks a lot for fixing this!

**Pull request merged**